app/lib/foreman_deployments/tasks/search_task_definition.rb

```diff
@@ -7,6 +7,26 @@ def run
           results = SerializableArray.new(results)
           SearchTaskDefinition.create_output(results, output)
         end
+
+        def self.dereference_output(storage, path)
```

Comment:
> No, it's still `plan CreateHost -> plan DhcpReserve -> execute CreateHost -> execute DhcpReserve`
It's not intended to do much with the inputs in the plan phase: the plan phase
is really mainly for planning and validation if possible.

In this case, DhcpReserve's `#plan` method will get a TaskReference object as a parameter. It's a bit non-trivial for the implementers, but as long as it's well documented, I'm OK with it.

> getting rid of the output sub-chains in planning is no-go here, as it's already being used by Dynflow users.

IMHO we can use them both - we will just use different access points to the output. For example we can leave the `#output` method as is (with chain recording) and create another method `#original_output` that will return the deserialized version of the object (without the chaining).

> I'm ok with introducing alias (but keeping the sub-key chaining)

That's basically what my patch does. If we move the implementation of `self.dereference_output` to `Actions::Base` class, it will become the default behavior, but still be overridable, if the task wishes to. Probably I will break it to two methods - one that deals with root deserializing and one for dealing with call chains. Then we would be able to get rid of the ugly `if` that I add to `ExecutionPlan::OutputReference#dereference`.
Basically we will just move the deserialization responsibility to the task itself - which anyway sounds reasonable to me.
What do you say @iNecas, does the last paragraph sound like a good starting point?